### DISCUSSION TITLE
read sample_rate, channels, bits_per_sample from unknown .wav file

### DISCUSSION THREAD

**Discussion by wegscheider** [pschatzmann/arduino-audio-tools#673] [Q&A] [answered]
Hello,

I would like to position playback of an .wav file on SD to a specific time. The position in terms of bytes can be easily converted if sample_rate, channels, bits_per_sample are known. I am aware of the setNotifyAudioChange function which allows me to properly output different .wav files, but how can I access these data directly in my code?

Thanks for your help and best regards,

Werner

**Comment by pschatzmann** [comment, marked as answer]
Why don't you just use the file api and do a lseek to the required position...

**Reply by wegscheider** [reply]
Yes, sure that's what I am doing, but depending on e.g. number of channels this position in bytes will correspond to a different time. 
I would need to know these three values to calculate the file position, e.g. fileposition = sample_rate*channels*bits_per_sample*time_in_secs...
In all examples I looked at so far these values are given and not taken from an unknown .wav file.

**Comment by pschatzmann** [comment]
Just write the header and then this information will be available. I suggest that you have a look at the WavDecoder documentation
and if you use an EncodedAudioStream this information gets updated automatically from the decoder...

**Reply by wegscheider** [reply]
Dear Phil:

Please excuse my ignorance, but I simply don't get it. I used mainly your code for playing .wav files (see below). This works fine, but if my .wav file is only 1 channel, I move forward to 2 minutes instead of 1. 

#include <SPI.h>
#include <SD.h>
#include "AudioTools.h"

#define SD_CS     5            
#define SD_MOSI   23
#define SD_MISO   19
#define SD_SCK    18

SPIClass SPISD(HSPI); // Creates SPI object using the HSPI interface

unsigned long runtime;
bool first = true;

I2SStream i2s; // final output of decoded stream
File audioFile;
EncodedAudioStream decoder(&i2s, new WAVDecoder()); // Decoding stream
StreamCopy copier(decoder, audioFile); 

void setup(){
  Serial.begin(115200);
  AudioLogger::instance().begin(Serial, AudioLogger::Info);  

  SPISD.begin(SD_SCK, SD_MISO, SD_MOSI); // Begin SPI used for SD card, not really necessary

  // setup file
  if(!SD.begin(SD_CS, SPISD)) {      
    Serial.println("SPI FAILED");
  }
 audioFile = SD.open("/HarlemSwing.wav");

  // setup i2s
  auto config = i2s.defaultConfig(TX_MODE);
  config.pin_bck = 14;
  config.pin_ws = 15;
  config.pin_data = 22;
  i2s.begin(config);

  // setup I2S based on sampling rate provided by decoder
  decoder.setNotifyAudioChange(i2s);
  decoder.begin();

  // begin copy
  copier.begin();

}

void loop(){
  runtime = millis();
   if (runtime > 5000) { // example on how file position can be changed to 1 minute
     if (first) {
       first = false;
// in order to get the right timing I would need to know the current sample_rate, channels, bits_per_sample
       audioFile.seek(44100*60*2*2); // works for 2 channel 16 bit, 44100 sampling rate
     }      
   }
  if (!copier.copy()) {
    audioFile.close();
    stop();
  }
}

**Comment by pschatzmann** [comment]
After the first copy the EncodedAudioStream knows this information and you can get it with decoder.audioInfo(). So you could add you first copier.copy(); to the setup, so that you have this available in the loop.

If this is too crude you can replace the copy with your own logic of reading from the file and writing to the destination. The wav header has 44 bytes, so that's the min amount of data that you need to write to get this information.

**Reply by wegscheider** [reply]
Great, works as you described. Thank's a lot.
